Processing/src/cz/fidentis/processing/decimation/QuadricEdgeCollapse.java

```diff
@@ -0,0 +1,433 @@
+                    to.sub(from);
+                    if(to.length() < threshold) {
+                        Edge e = new Edge(i, j);
+                        if(!pairs.contains(e)) {
+                            pairs.add(e);
+                        }
+                    }
+                }
+            }
+        }
+    }
+    
+    /**
+     * For each pair considered for contraction we need to determine the position
+     * of contracted vertex.
+     */
+    private void initializeTargets() {
+        pairTargets = new HashMap<>(pairs.size());
+        
+        for(Edge e : pairs) {
+            pairTargets.put(e, computeTargetV(e));
+        }
+    }
+    
+    /**
+     * Finally, initialised pairs should be sorted by the cost of contraction so
+     * that the least errors would be produced.
+     */
+    private void initializeQueue() {
+        pairErrors = new HashMap<>(pairs.size());
+        EdgeComparator c = new EdgeComparator(pairErrors);
+        queue = new PriorityQueue<>(pairs.size(), c);
+        
+        for(Edge e : pairs) {
+            // get weight of this pair
+            double weight = computeWeight(e);
+            pairErrors.put(e, weight);
+            
+            // add it to priority queue keyed by weight
+            queue.add(e);
+        }
+    }
+    
+    /**
+     * Removes all vertices and faces that were detached during collapses. Also
+     * recompute normals.
+     */
+    private void cleanupModel() {
+        // TODO do this
+        ArrayList<Vector3f> newVerts = new ArrayList<>();
+        
+        for(int i=0;i<model.getVerts().size();i++) {
+            if(!removedVerts.contains(i)) {
+                newVerts.add(model.getVerts().get(i));
+            }
+        }
+    }
+    
+    /**
+     * Gets the edge that adds the least significant error if collapsed, according
+     * to the priority queue keyed by error quadric of edge.
+     * @return 
+     */
+    private Edge getEdgeToContract() {
+        return queue.poll();
+    }
+    
+    /**
+     * Contracts given edge to one vertex on a position which creates the least
+     * error. Then updates mesh, error quadrics and weights accordingly.
+     * @param e 
+     */
+    private void contractEdge(Edge e) {
+        System.out.println("Starting edge " + e.getFrom() + " to " + e.getTo());
+        int fromIdx = getModifiedVertexIndex(e.getFrom());
+        int toIdx = getModifiedVertexIndex(e.getTo());
+        
+        // move TO to the target collapsed location
+        Vector3f target = getTargetV(e);
+        Vector3f toVertex = model.getVerts().get(toIdx);
+        toVertex.x = target.x;
+        toVertex.y = target.y;
+        toVertex.z = target.z;
+        
+        // reattach all triangles adjacent to FROM to the moved TO
+        // FROM should be removed, with faces that would degenerate to 2D
+        // TODO: actually remove correct faces and vertices
```

Comment:
No, ono tam uz nema byt, odstranovanie toho co netreba sa robi na konci (takze sa to naozaj robi)